### PR TITLE
fix(app): fail-closed sanctions guard + gate localStorage RPC overrides behind debug flag

### LINKED ISSUE #397: fix(app): fail-closed sanctions guard + gate localStorage RPC overrides behind debug flag
Part of the 2026-07 repository audit — M1 — Critical correctness & security.

Two client-side security gaps, both small, fixed in one PR. **Part A:** the sanctions API route fails closed (502 on any error), but the client guard only blocks when `isSanctioned === true` — so a failed or blocked API call lets a sanctioned user straight into the app, and children render while the check is still in flight. **Part B:** `packages/web3/src/config/chains.ts` honours `localStorage` RPC-override keys unconditionally in production, so anything that can write `localStorage` (XSS, a malicious extension, "paste this in your console" self-XSS) can persistently repoint mainnet reads and swap quotes to an attacker-controlled RPC.

Be honest about scope: client-side sanctions screening is inherently bypassable by a determined user editing JS in DevTools. Part A makes the *default* path fail closed, matching the server route's explicit design intent. Server-side enforcement is out of scope here.

### Part A — sanctions guard fails open on the client

The server route fails closed. `apps/app.mento.org/app/api/sanctions/route.ts:57-62`:

```ts
function failClosed() {
  return NextResponse.json(
    { isSanctioned: null, error: "check_failed" },
    { status: 502 },
  );
}
```

`failClosed()` is returned on missing API key (route.ts:80-83), non-OK upstream response (line 107), malformed payload (line 122), and fetch errors (line 139).

The client hook already computes and returns everything the guard needs. `apps/app.mento.org/app/hooks/use-sanctions-check.ts:51-55` and `:79`:

```ts
const isSanctioned =
  data?.isSanctioned === true ||
  (blockedAddress !== null && (!address || address === blockedAddress));
const checkFailed = isError && !isLoading;
const isChecking = isLoading && isConnected && !!address;
// ...
return { isSanctioned, isChecking, checkFailed };
```

But the guard consumes only `isSanctioned`. `apps/app.mento.org/app/components/sanctions-guard.tsx:7-10` and `:28`:

```tsx
export function SanctionsGuard({ children }: PropsWithChildren) {
  const { isSanctioned } = useSanctionsCheck();

  if (isSanctioned) {
    // ... blocked screen ...
  }
  return children;
}
```

Consequences:

1. If `/api/sanctions` returns 502 (or the user blocks the request in DevTools), the query errors, `isSanctioned` stays `false`, and the app renders normally. The server's fail-closed design is undone on the client.
2. While the check is in flight (`isChecking === true`), `children` render — a sanctioned user has a working app for the duration of the round-trip.

The guard wraps the whole app: `apps/app.mento.org/app/components/providers.tsx:41-43`.

There is also a misleadingly named existing test — it mocks only `isSanctioned: false` and never exercises an in-flight state. `apps/app.mento.org/app/components/sanctions-guard.test.tsx:38-44`:

```ts
it("renders children while checking", () => {
  useSanctionsCheckMock.mockReturnValue({ isSanctioned: false });

  renderGuard();

  expect(screen.queryByTestId("app-content")).not.toBeNull();
});
```

### Part B — localStorage RPC overrides honoured in production

`packages/web3/src/config/chains.ts` reads three kinds of `localStorage` keys unconditionally, in every environment:

Fork-mode flag — chains.ts:258-277 (read at line 265; note the module-level `const useFork = isForkModeEnabled();` at line 17):

```ts
function isForkModeEnabled(): boolean {
  if (typeof window === "undefined") {
    // During SSR, check environment variable
    return process.env.NEXT_PUBLIC_USE_FORK === "true";
  }

  // In browser, check localStorage first, then environment variable
  const storedValue = localStorage.getItem("mento_use_fork");
  if (storedValue !== null) {
    return storedValue === "true";
  }
```

Legacy global override — chains.ts:279-285 (read at line 281):

```ts
function getLegacyCustomRpcUrl(): string | undefined {
  if (typeof window !== "undefined") {
    const stored = localStorage.getItem("mento_custom_rpc_url");
    if (stored) return stored;
  }
  return process.env.NEXT_PUBLIC_RPC_URL || undefined;
}
```

Per-chain overrides — chains.ts:287-300 (read at line 293; keys like `mento_custom_rpc_url_42220` defined in `RPC_OVERRIDE_CONFIG`, chains.ts:28-62):

```ts
function getChainSpecificRpcUrl(
  chainId: ChainId,
): { url: string; source: string } | undefined {
  const config = RPC_OVERRIDE_CONFIG[chainId];

  if (typeof window !== "undefined") {
    const stored = localStorage.getItem(config.localStorageKey);
    if (stored) {
      return {
        url: stored,
        source: `custom RPC (${config.localStorageKey})`,
      };
    }
  }
```

Environment helpers already exist. `packages/web3/src/utils/environment.ts:1-3` (re-exported via `packages/web3/src/utils/index.ts:7`):

```ts
export const IS_PROD = process.env.NEXT_PUBLIC_VERCEL_ENV === "production";
export const IS_DEV = process.env.NEXT_PUBLIC_VERCEL_ENV === "development";
export const IS_DEBUG = process.env.NEXT_PUBLIC_ENABLE_DEBUG === "true";
```

The legitimate dev/fork workflow this must not break: root `package.json:15-16` (`"fork:mainnet": "anvil --fork-url https://forno.celo.org --port 8545"`, plus `fork:testnet`) combined with `NEXT_PUBLIC_USE_FORK=true` — that path goes through the env-var branches (chains.ts:261 and :271-273), not `localStorage`. All three env vars are already declared in `turbo.json` (lines 10-11, 22).

### Part A (app.mento.org)

1. Modify `apps/app.mento.org/app/components/sanctions-guard.tsx`: consume all three flags and branch — sketch, adapt to local style:

   ```tsx
   const { isSanctioned, isChecking, checkFailed } = useSanctionsCheck();

   if (isSanctioned || checkFailed) {
     return ( /* existing blocked-screen layout */ );
   }

   if (isChecking) {
     return ( /* neutral full-screen loading state — NOT {children} */ );
   }

   return children;
   ```

   - For `checkFailed && !isSanctioned`, vary the copy inside the same blocked layout: heading like "Verification unavailable", body explaining the compliance check could not be completed and to retry later. Do not claim the address is sanctioned when the check merely failed.
   - The loading state should be a simple centred placeholder (match the blocked screen's container classes: `min-h-screen`, `bg-background`). No spinner library — a static element or a Tailwind `animate-pulse` div is fine.
2. Update `apps/app.mento.org/app/components/sanctions-guard.test.tsx`. Read all three existing test files first (`route.test.ts`, `sanctions-guard.test.tsx`, `use-sanctions-check.test.tsx`) — the hook and route are already well covered; the new cases belong in the guard test:
   - Replace the misleading test at lines 38-44 ("renders children while checking") with one that mocks `{ isSanctioned: false, isChecking: true, checkFailed: false }` and asserts children are NOT rendered and the blocked copy is NOT shown.
   - Add: `{ isSanctioned: false, isChecking: false, checkFailed: true }` → blocked UI rendered, children not rendered.
   - Update existing mocks to return all three flags (the current mocks return only `isSanctioned`, which would make `isChecking`/`checkFailed` `undefined` — falsy, so passing, but make them explicit).
3. No changes to `route.ts`, `route.test.ts`, or `use-sanctions-check.ts` are expected — the hook already exposes the flags and the route already fails closed.

### Part B (packages/web3)

4. Create `packages/web3/src/config/rpc-overrides.ts` — a small pure, dependency-injected module so it is testable under the node vitest environment (sketch — adapt to local style):

   ```ts
   import { IS_DEBUG, IS_PROD } from "../utils/environment";

   export type StorageLike = Pick<Storage, "getItem">;

   /** localStorage RPC/fork overrides are honoured only outside production
    *  builds, or when the debug flag was baked in at build time. */
   export function canUseStorageOverrides(
     isProduction: boolean,
     isDebugEnabled: boolean,
   ): boolean {
     return isDebugEnabled || !isProduction;
   }

   export function readStorageOverride(
     key: string,
     storage: StorageLike | undefined = typeof window === "undefined"
       ? undefined
       : window.localStorage,
     isProduction: boolean = IS_PROD,
     isDebugEnabled: boolean = IS_DEBUG,
   ): string | null {
     if (!storage) return null;
     if (!canUseStorageOverrides(isProduction, isDebugEnabled)) return null;
     return storage.getItem(key);
   }
   ```

5. In `packages/web3/src/config/chains.ts`, route all three `localStorage.getItem` calls through the helper:
   - chains.ts:265 → `const storedValue = readStorageOverride("mento_use_fork");`
   - chains.ts:281 → `const stored = readStorageOverride("mento_custom_rpc_url");`
   - chains.ts:293 → `const stored = readStorageOverride(config.localStorageKey);`

   Leave every env-var branch untouched: `NEXT_PUBLIC_USE_FORK` (lines 259-262, 271-273), `config.envUrl` (lines 302-308), `NEXT_PUBLIC_RPC_URL` (line 284). The fork workflow (`pnpm fork:mainnet` + `NEXT_PUBLIC_USE_FORK=true`) must keep working unchanged.
6. Create `packages/web3/src/config/rpc-overrides.test.ts`, following the plain describe/it style of `packages/web3/src/config/chain-policy.test.ts`. Cover at least:
   - `canUseStorageOverrides`: (prod, no debug) → false; (prod, debug) → true; (non-prod, no debug) → true.
   - `readStorageOverride` with an injected fake `{ getItem }`: returns the stored value when allowed; returns `null` without calling `getItem` when prod + no debug; returns `null` when storage is `undefined` (SSR).

## Out of scope

- Server-side/middleware sanctions enforcement (making the check unbypassable). This PR only makes the default client path fail closed.
- Rate-limiter or Chainalysis changes in `route.ts` — the route already fails closed.
- Removing the fork/override feature itself, changing `RPC_OVERRIDE_CONFIG`, or touching the `NEXT_PUBLIC_*` env-var override paths.
- The stale CLAUDE.md docs (React Aria/onPress claim) — separate issue.
- Any new npm dependencies (none are needed; none are approved).

## Acceptance criteria

- [ ] `SanctionsGuard` renders the blocked screen when `checkFailed` is true, even with `isSanctioned` false.
- [ ] `SanctionsGuard` renders a neutral loading state (children NOT mounted) while `isChecking` is true.
- [ ] Clean connected address and unconnected visitors still see `children` (existing tests at `sanctions-guard.test.tsx` stay green).
- [ ] `grep -n "localStorage.getItem" packages/web3/src/config/chains.ts` returns no matches (all reads go through `readStorageOverride`).
- [ ] `packages/web3/src/config/rpc-overrides.test.ts` exists and covers the prod-blocked, debug-allowed, non-prod-allowed, and SSR cases.
- [ ] `turbo test --filter app.mento.org` passes.
- [ ] `turbo test --filter @repo/web3` passes.
- [ ] `turbo check-types` passes.
- [ ] `trunk check --fix` reports no new issues.
- [ ] Runtime check (dev keeps working): `turbo dev --filter app.mento.org`, then in the browser console `localStorage.setItem("mento_custom_rpc_url_42220", "http://localhost:9999")` and reload — with no `NEXT_PUBLIC_VERCEL_ENV` set (non-production), Celo RPC requests go to the override (visible as failed requests to `localhost:9999` in the Network tab). Remove the key afterwards.
- [ ] Production gating check: `NEXT_PUBLIC_VERCEL_ENV=production turbo build --filter app.mento.org` succeeds; with the same localStorage key set against `next start`, RPC requests go to `forno.celo.org`, not the override.

## Gotchas

- **`NEXT_PUBLIC_*` inlining:** these vars are string-replaced at build time. They only work as static literal member expressions (`process.env.NEXT_PUBLIC_ENABLE_DEBUG`) — never construct the name dynamically. `utils/environment.ts` already does this correctly; import `IS_PROD`/`IS_DEBUG` from there rather than re-reading `process.env` in new code.
- **Module-load evaluation:** `chains.ts` computes `const useFork = isForkModeEnabled()` at line 17 and builds all chain objects at module scope. The gate must live inside the read functions; do not attempt to make it reactive or to add a `chains.test.ts` (module-level side effects and `.svg` imports make it hostile to the node test environment — that is exactly why the helper is extracted).
- **web3 tests run in node, not jsdom:** `packages/vitest-config/vitest.shared.js` sets `environment: "node"` and `packages/web3/vitest.config.ts` does not override it. There is no `window`/`localStorage` in these tests — inject the fake `StorageLike`; do not `vi.stubGlobal("localStorage", ...)` and expect the `window` check to pass.
- **SSR:** `chains.ts` is imported during server rendering. The helper must return `null` when `window` is undefined (the sketch handles this via the default parameter).
- **`mento_use_fork` semantics:** today a stored `"false"` explicitly disables fork mode even when `NEXT_PUBLIC_USE_FORK=true` (chains.ts:266-268). Gating by returning `null` from the helper means production ignores the key entirely and falls through to the env var — that is the intended behaviour.
- **Guard branch order:** check `isSanctioned || checkFailed` before `isChecking`. During the in-flight window `checkFailed` is false by construction (`isError && !isLoading`), so there is no flash of the blocked screen while loading.
- **No global spinner for visitors:** `isChecking` is only true when `isConnected && !!address` (use-sanctions-check.ts:55). Unconnected users never see the loading state — do not "improve" this by gating on `isLoading` alone.
- **react-query retries:** the hook uses `retry: 2`, so `checkFailed` only turns true after three failed attempts. Existing hook tests use `waitFor(..., { timeout: 5000 })` for this; guard tests avoid the problem entirely by mocking the hook (keep that pattern).
- **Existing guard test mocks return partial shapes:** after the change, `useSanctionsCheckMock` return values should include all three flags explicitly, or TypeScript may not catch a renamed flag (the mock is untyped via `vi.fn()`).
- **Honesty in copy:** a failed check is not a sanctions hit. Use distinct copy for `checkFailed` so support tickets are triageable.

**Effort:** M · **Risk:** low

### PR DESCRIPTION
**Pull request by chapati23**
Closes #397

## What / Why

Two client-side security gaps, fixed together per the issue.

**Part A** — `SanctionsGuard` only blocked on `isSanctioned === true`. A failed or blocked `/api/sanctions` call left `isSanctioned` false, so a failed check let a potentially sanctioned user straight into the app, and `children` rendered while the check was still in flight. The guard now also blocks on `checkFailed` (with distinct "Verification unavailable" copy — a failed check is not a sanctions hit) and shows a neutral loading placeholder (no `children`) while `isChecking` is true.

**Part B** — `packages/web3/src/config/chains.ts` read `localStorage` RPC/fork override keys unconditionally in every environment, so anything that can write `localStorage` (XSS, malicious extension, self-XSS) could persistently repoint mainnet RPC reads and swap quotes. Extracted a small, dependency-injected `packages/web3/src/config/rpc-overrides.ts` (`canUseStorageOverrides` / `readStorageOverride`) that only honours the override when not in production, or when `NEXT_PUBLIC_ENABLE_DEBUG` was baked in at build time. All three `localStorage.getItem` call sites in `chains.ts` now route through it; the `NEXT_PUBLIC_*` env-var override paths (including the `fork:mainnet`/`NEXT_PUBLIC_USE_FORK` workflow) are untouched.

## Verification

```
pnpm --filter @repo/web3 exec tsc --noEmit         # no output — passes
pnpm --filter app.mento.org exec tsc --noEmit      # no output — passes
pnpm --filter @repo/web3 build                     # rebuilt dist (source changed)
pnpm --filter @repo/web3 exec vitest run           # 11 files, 79 tests passed
pnpm --filter app.mento.org exec vitest run        # 19 files, 104 tests passed
grep -n "localStorage.getItem" packages/web3/src/config/chains.ts   # no matches
pnpm --filter app.mento.org exec eslint app/components/sanctions-guard.tsx app/components/sanctions-guard.test.tsx   # clean
pnpm --filter @repo/web3 exec eslint src/config/chains.ts src/config/rpc-overrides.ts src/config/rpc-overrides.test.ts   # clean
pnpm exec prettier --write <changed files>         # formatting fixed, re-ran affected tests, still green
git commit (pre-commit hook ran trunk check --fix on staged files)   # "No issues"
```

Note: `npx @trunkio/launcher fmt` itself failed locally with an `EPERM` on the shared npm cache (root-owned `~/.npm/_cacache`), unrelated to this change — used `prettier --write` + per-package `eslint` directly instead, and the commit's pre-commit hook (which does run trunk) reported no issues.

## Deferred (human step)

The issue's manual runtime checks (browser console `localStorage.setItem` against `turbo dev`, and the `NEXT_PUBLIC_VERCEL_ENV=production turbo build` + `next start` gating check) were not run in this sandboxed session — logic is covered by the new `rpc-overrides.test.ts` unit tests instead.

---

> [!NOTE]
> **High Risk**
> Touches compliance access control and mainnet RPC selection—both security-sensitive; behavior changes are intentional fail-closed fixes.
> 
> **Overview**
> **SanctionsGuard** now fails closed: app content stays hidden while `isChecking`, blocks on `checkFailed` with *Verification unavailable* (distinct from a sanctions hit), and only renders children after a clean, completed check.
> 
> **RPC/fork `localStorage` overrides** in `@repo/web3` are routed through new `readStorageOverride` / `canUseStorageOverrides` helpers so production builds ignore those keys unless the debug flag was baked in at build time; env-based RPC/fork paths are unchanged. Unit tests cover the gating and storage edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9db268041552a0e3cc49a81db4be045d620d3509. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/web3/src/config/rpc-overrides.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  isDebugEnabled: boolean,
+): boolean {
+  return isDebugEnabled || !isProduction;
+}
+
+export function readStorageOverride(
+  key: string,
+  storage: StorageLike | undefined = typeof window === "undefined"
+    ? undefined
+    : window.localStorage,
```
[the remark this answers is not in the document]
Comment (chapati23) [reply packages/web3/src/config/rpc-overrides.ts]: Fixed in 9db2680 — production builds now check the debug gate before touching window.localStorage, and blocked storage access returns no override instead of crashing.
